File: sdk/typescript/src/paths.ts
import {
  collectEnumsNative,
  configurePathConstantsNative,
  getRootToolKeyNative,
  pathBuilderMemoryOnlyNative,
  pathCatalogPrefixNative,
  pathDefaultCatalogDirNative,
  pathDecomposedPrefixNative,
  pathDecomposedRootNative,
  pathJsonExtNative,
  pathMdExtNative,
  pathWriteCatalogPruneNative,
  toDecomposedKeyNative,
  toolIdFromDecomposedRelNative,
} from "./native.js";
import type { JsonRecord } from "./types.js";

/** Push host app overrides into native PathConfig (Rust defaults when not called). */
export function configurePathConstants(opts: {
  mdExt: string;
  jsonExt: string;
  decomposedPrefix: string;
  decomposedRoot: string;
  catalogPrefix: string;
  builderMemoryOnly: boolean;
  defaultCatalogDir: string;
  writeCatalogPrune: boolean;
}): void {
  configurePathConstantsNative({
    mdExt: opts.mdExt,
    jsonExt: opts.jsonExt,
    decomposedPrefix: opts.decomposedPrefix,
    decomposedRoot: opts.decomposedRoot,
    catalogPrefix: opts.catalogPrefix,
    builderMemoryOnly: opts.builderMemoryOnly,
    defaultCatalogDir: opts.defaultCatalogDir,
    writeCatalogPrune: opts.writeCatalogPrune,
  });
}

export function mdExt(): string {
  return pathMdExtNative();
}

export function jsonExt(): string {
  return pathJsonExtNative();
}

export function decomposedPrefix(): string {
  return pathDecomposedPrefixNative();
}

export function decomposedRoot(): string {
  return pathDecomposedRootNative();
}

export function catalogPrefix(): string {
  return pathCatalogPrefixNative();
}

export function defaultCatalogDir(): string {
  return pathDefaultCatalogDirNative();
}

export function builderMemoryOnly(): boolean {
  return pathBuilderMemoryOnlyNative();
}

export function writeCatalogPrune(): boolean {
  return pathWriteCatalogPruneNative();
}

export function toDecomposedKey(filePath: string): string | null {
  return toDecomposedKeyNative(filePath);
}

export function toolIdFromDecomposedRel(relPath: string): string {
  return toolIdFromDecomposedRelNative(relPath);
}

export function getRootToolKey(filePath: string): string | null {
  return getRootToolKeyNative(filePath);
}

export function collectEnums(schema: unknown): unknown[] {
  return collectEnumsNative(schema as JsonRecord);
}

File: sdk/go/paths.go
package cytindexer

// CollectEnums extracts enum values from a JSON schema.
func CollectEnums(schemaJSON string) (string, error) {
	return cgoCollectEnums(schemaJSON)
}

// ToDecomposedKey converts a file path to a decomposed catalog key.
func ToDecomposedKey(filePath string) (string, error) {
	return cgoToDecomposedKey(filePath)
}

// ToolIDFromDecomposedRel derives a tool id from a decomposed relative path.
func ToolIDFromDecomposedRel(relPath string) (string, error) {
	return cgoToolIDFromDecomposedRel(relPath)
}

// GetRootToolKey returns the root tool key for a decomposed file path.
func GetRootToolKey(filePath string) (string, error) {
	return cgoGetRootToolKey(filePath)
}

// ConfigurePathConstants overrides native path configuration constants.
func ConfigurePathConstants(mdExt, jsonExt, decomposedPrefix, decomposedRoot, catalogPrefix, defaultCatalogDir string, builderMemoryOnly, writeCatalogPrune bool) error {
	return cgoConfigurePathConstants(mdExt, jsonExt, decomposedPrefix, decomposedRoot, catalogPrefix, defaultCatalogDir, builderMemoryOnly, writeCatalogPrune)
}

// PathMdExt returns the configured markdown file extension.
func PathMdExt() (string, error) {
	return cgoPathMdExt()
}

// PathJsonExt returns the configured JSON file extension.
func PathJsonExt() (string, error) {
	return cgoPathJsonExt()
}

// PathDecomposedPrefix returns the decomposed catalog path prefix.
func PathDecomposedPrefix() (string, error) {
	return cgoPathDecomposedPrefix()
}

// PathDecomposedRoot returns the decomposed catalog root directory.
func PathDecomposedRoot() (string, error) {
	return cgoPathDecomposedRoot()
}

// PathCatalogPrefix returns the catalog path prefix.
func PathCatalogPrefix() (string, error) {
	return cgoPathCatalogPrefix()
}

// PathDefaultCatalogDir returns the default catalog directory.
func PathDefaultCatalogDir() (string, error) {
	return cgoPathDefaultCatalogDir()
}

// PathBuilderMemoryOnly reports whether the catalog builder defaults to memory-only mode.
func PathBuilderMemoryOnly() (bool, error) {
	return cgoPathBuilderMemoryOnly()
}

// PathWriteCatalogPrune reports whether catalog writes include prune metadata.
func PathWriteCatalogPrune() (bool, error) {
	return cgoPathWriteCatalogPrune()
}
